Added Simulated Annealing

Proxihox opened the pull request:
Article with explanation of approach for simulated annealing and a template code which can be used to implement the same.

Closes #996

mhayter commented:
I haven't thoroughly reviewed yet, but initially what I see is that README.md needs to be updated and some of the formatting is not rendering properly (reduce the spaces between $$). 
You can preview: https://cp-algorithms.com/preview.html. 

Please give the title of the practice problem in addition to the links.(Hyperlink the title of the problem). 

Consider naming the function and making it callable rather than using ```int main{}```

Thanks.

Proxihox commented:
@mhayter , Thanks for reviewing my PR. I've fixed the changes you mentioned. I'm not able to figure out why the preview throws an error when I write the following cpp inline code : 
```cpp
points = {{0,0},{2,2},{0,2},{2,0},{0,1},{1,2},{2,1},{1,0}};
```
This is trivial, as the actual values of these points isnt important and for now I've replaced it with a comment asking the user to fill it in with a set of points of their own choice. 
 Lmk if there is anything else I need to change.

mhayter commented:
Thanks for your progress this far. Unfortunately there are still many variables/ formula that don't render well or are not formatted at all like  $s$ , $s_next$,$T$, the acceptance function, etc. 

Substance wise, I believe that the analogy of how 'temperature' plays into this process need to be explained likely via analogy.

Also, it may make more sense to have a more descriptive function name than $E()$. 

Again these are just initial thoughts. I hoping someone more senior, will review this as well but we we owe you prompt feedback given your timely response.

jxu commented:
I've implemented simulating annealing for fun in the past but not for programming contests because randomized and not guaranteed to get the right answer. Unlike say convex optimization or iterative relaxation which is guaranteed to converge.

Proxihox commented:
@mhayter Thanks once more for the quick response. I agree that the temperature part could have been elaborated more, so I added a small section on how this algorithm works. with respect to the formatting, I've tried to fix everything, but I'll get a friend to cross verify it once more.

mhayter commented:
Okay, I've been going through the code more closely.  

I've implemented a simple dp (though I probably should have used chatgpt or google) to solve TSP and paired it against this algorithm. 

- As far as the algorithms is concerned, its not clear to me as a novice how I should set `T` and `u`. 

- Stylistically, we can use `swap()` in c++ for the points indices.
- Choosing the points indices is quite strange and should cause overflow (rand() % points.size() should roughly work)
- We need to make it more obvious on how to initialized to points array:
    (Perhaps consider having an copy constructor i.e. ```state(const vector<pair<int,int>> &points) ```)?

Removing capitalization of `state` would be consistent.  

Cool article though. I was able to get a prototype working which I haven't done before..

I'd still like @adamant-pwn or @jakobkogler  to review.

Proxihox commented:
@mhayter Thanks for suggesting the improvements.
- I wrote how I'd approach choosing values for `T` and `u`.
- Initially, I didn't use `exp()` , because I wanted to show that it is possible to modify the base of the exponent. I've explained why this is useful in the further modification section. I still modified the original PAF to use `exp()`.
- I implemented the copy constructor in the format `state(state& s);`, to keep up with the definition of the copy constructor.

I also figured out the cause for the double curly braces problem from my first commit : [Double curly braces and HTML](https://stackoverflow.com/questions/17878560/difference-between-double-and-single-curly-brace-in-angular-js)
and how to fix it : [Solution to the double curly braces problem](
https://stackoverflow.com/questions/24102498/escaping-double-curly-braces-inside-a-markdown-code-block-in-jekyll)

 Let me know if there are any further changes.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+
+## The problem
+
+We are given a function $E(s)$, which calculates the potential of the state $s$. We are tasked with finding the state $s_{best}$ at which $E(s)$ is minimized. **SA** is suited for problems where the states are discrete and $E(s)$ has multiple local minima. We'll take the example of the [Travelling Salesman Problem (TSP)](https://en.wikipedia.org/wiki/Travelling_salesman_problem). 
```
> which calculates the potential of the state $s$

Please define "potential" before using the term.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+
+### State
+
+State space is the collection of all possible values that can be taken by the independent variables.
```
> independent variables

Independent from what and in which sense? This also uses the term without defining it. Also wouldn't it be easier to say that the state space is the domain of the energy function, and a state is any its element?

Proxihox replied:
Yup went ahead with your suggestion for this.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+
+It is a state in the state space which is close to the previous state. This usually means that we can obtain the neighbouring state from the original state using a simple transform. In the case of the Travelling Salesman Problem, a neighbouring state is obtained by randomly choosing 2 nodes, and swapping their positions in the current state. 
+
+### The energy function E(s)
```
Again, the term energy is used without a definition or any kind of informal explanation as to why this specific term is applied to $E(s)$ in particular.

Proxihox replied:
Its analogous to the internal energy of the material undergoing annealing, which is explained in the Motivation section.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+
+### How does this work?
+
+This algorithm is called simulated annealing because we are simulating the process of annealing, wherein a material is heated up and allowed to cool, in order to allow the atoms inside to rearrange themselves in an arrangement with minimal internal energy, which in turn causes the material to have different properties. The state is the arrangement of atoms and the internal energy is the function  ...
```
I would move this explanation much closer to the beginning of the article.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+
+This algorithm is called simulated annealing because we are simulating the process of annealing, wherein a material is heated up and allowed to cool, in order to allow the atoms inside to rearrange themselves in an arrangement with minimal internal energy, which in turn causes the material to have different properties. The state is the arrangement of atoms and the internal energy is the function  ...
+
+Simulated annealing, literally simulates this process. We start off with some random state (material) and set a high temperature (heat it up). Now, the algorithm is ready to accept states which have a higher energy than the current state, as it is motivated by the high value of $T$. This prevents the algorithm from getting stuck inside local minimas and move towards the global minima. As time pro ...
```
```suggestion
Simulated annealing, literally, simulates this process. We start off with some random state (material) and set a high temperature (heat it up). Now, the algorithm is ready to accept states which have a higher energy than the current state, as it is motivated by the high value of $T$. This prevents the algorithm from getting stuck inside local minimas and move towards the global minima. As time progresses, the algorithm cools down and refuses the states with higher energy and moves into the closest minima it has found.
```

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+<br>
+<i>A visual representation of simulated annealing, searching for the maxima of this function with multiple local maxima.</i>.
+<br>
+<i>This <a href="https://upload.wikimedia.org/wikipedia/commons/d/d5/Hill_Climbing_with_Simulated_Annealing.gif">gif</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
```
```suggestion
<i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
```
Note that, in principle, CC0 doesn't require any attribution (but we can still keep it as a token of respect).

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+
+## Probability Acceptance Function(PAF)
+
+$P(E,E_{next},T) = 
+    \begin{cases}
+       \text{True} &\quad\text{if }  \exp{\frac{-(E_{next}-E)}{T}} \ge random(0,1) \\
+       \text{False} &\quad\text{otherwise}\\
+     \end{cases}$
+
+This function takes in the current state, the next state and the Temperature , returning a boolean value, which tells our search whether it should move to $s_{next}$ or stay at $s$. Note that for $E_{next} < E$ , this function will always return True. 
```
```suggestion
$P(E,E_{next},T) = 
    \begin{cases}
       \text{True} &\quad\text{if }  \mathcal{U}_{[0,1]} \le \exp(-\frac{E_{next}-E}{T}) \\
       \text{False} &\quad\text{otherwise}\\
     \end{cases}$

Here, $\mathcal{U}_{[0,1]}$ is a continuous uniform random value on $[0,1]$. This function takes in the current state, the next state and the temperature, returning a boolean value, which tells our search whether it should move to $s_{next}$ or stay at $s$. Note that for $E_{next} < E$ , this function will always return True, otherwise it can still make the move with probability $\exp(-\frac{E_{next}-E}{T})$, which corresponds to the [Gibbs measure](https://en.wikipedia.org/wiki/Gibbs_measure).
```

adamant-pwn commented on src/num_methods/simulated_annealing.md:
````diff
@@ -0,0 +1,198 @@
+
+```cpp
+bool P(double E,double E_next,double T){
+    double prob = (double)rand()/RAND_MAX; // Generate a random number between 0 and 1
````
I think this naming is misleading. The actual transition probability is $\exp(-\frac{E_{next}-E}{T})$, and this value is only used to emulate it. So, ideally we want to rewrite it using e.g. `std::bernoulli_distribution`.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+        // Generate the initial state
+    }
+    state(state& s){
+        // Implement the copy constructor
+    }
+    state next(){
+        state s_next = state(*this);
+
+        // Modify s_next to the neighbouring state
+
+        return s_next;
+    }
+    double E(){
+        // Implement the cost function here
+    };
+};
+
+pair<double,state> simAnneal(){
+    state s = state();
+    state best = state(s);
+    double T = 1000; // Initial temperature
+    double u = 0.99; // decay rate
+    double E = s.E();
+    double E_next;
+    double E_best = E;
+    while (T > 1){
+        state next = s.next();
+        E_next = next.E();
+        if(P(E,E_next,T)){
+            s = next;
+            if(E_next < E_best){
+                best = s;
+                E_best = E_next;
+            }
+        }
+        E = E_next;
+        T *= u;
+    }
+    return {E_best,best};
+}
```
```suggestion
class state {
    public:
    state() {
        // Generate the initial state
    }
    state(state const& s) {
        // Implement the copy constructor
    }
    state next() {
        state s_next = state(*this);

        // Modify s_next to a random neighboring state

        return s_next;
    }
    double E() {
        // Implement the energy function here
    };
};

pair<double, state> simAnneal() {
    state s = state();
    state best = state(s);
    double T = 1000; // Initial temperature
    double u = 0.99; // decay rate
    double E = s.E();
    double E_next;
    double E_best = E;
    while (T > 1) {
        state next = s.next();
        E_next = next.E();
        if(P(E, E_next, T)) {
            s = next;
            if(E_next < E_best) {
                best = s;
                E_best = E_next;
            }
        }
        E = E_next;
        T *= u;
    }
    return {E_best, best};
}
```

adamant-pwn replied:
Actually I doubt you even need an explicit copy constructor, usually the default one is fine?

Proxihox replied:
It might be required for some problems, but it is a nuisance to keep it there by default so I've removed it.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+    state s = state();
+    state best = state(s);
+    double T = 1000; // Initial temperature
+    double u = 0.99; // decay rate
```
It seems to be the first time the "decay" is introduced. It should be mentioned earlier, as well as some explanation to why the rate of decay is exponential.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
````diff
@@ -0,0 +1,198 @@
+
+```
+## How to use:
+Fill in the state class functions as appropriate. If you are trying to find a global maxima and not a minima, ensure that the $E()$ function returns negative of the function you are maximising and finally print out $-E_{best}$. Set the below parameters as per your need.
````
```suggestion
Fill in the state class functions as appropriate. If you are trying to find a global maxima and not a minima, ensure that the $E()$ function returns negative of the function you are maximizing and print $-E_{best}$ in the end. Set the below parameters as per your need.
```

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+    }
+    state next(){
+        state s_next = state(*this);
+        int a = (rand()%points.size());
```
Using `rand()` is [not safe](https://codeforces.com/blog/entry/61587). It may be acceptable in some cases, but in this particular case, it can severely backfire e.g. because on some systems it's always a 16-bit number.

Proxihox replied:
Was an interesting read, implemented the Mersenne twister

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+    }
+
+    double euclidean(pair<int,int> a, pair<int,int> b){
+        return pow(pow((a.first-b.first),2)+pow((a.second-b.second),2),0.5);
```
```suggestion
        return std::hypot(a.first - b.first, a.second - b.second);
```

adamant-pwn replied:
Please also put space between `)` and `{`, after `,`, and around arithmetic operators.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+
+## Problems
+
+- [USACO Jan 2017 - Subsequence Reversal](https://usaco.org/index.php?page=viewproblem2&cpid=698)
+- [Deltix Summer 2021 - DIY Tree](https://codeforces.com/contest/1556/problem/H)
+- [AtCoder Contest Scheduling](https://atcoder.jp/contests/intro-heuristics/tasks/intro_heuristics_a)
```
Did you test the implementation from the article on any of those?

Proxihox replied:
I've tested the implementation with the 3rd problem : [AtCoder Contest Scheduling](https://atcoder.jp/contests/intro-heuristics/tasks/intro_heuristics_a).

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,198 @@
+## Further modifications to the algorithm:
+
+- Add a time based exit condition to the while loop to prevent TLE
+- The Probability acceptance function given above, prefers accepting states which are lower in energy because of the $E_{next} - E$ factor in the numerator of the exponent. You can simply remove this factor, to make the PAF independent of the difference in energies.
```
But why and when would one do it?

Proxihox replied:
Consider an E(s) which is pretty much constant for most points in its domain, but has very sharp drops for its minima's. In such a case, after finding a particular local minima, the PAF will mostly reject all points which are not inside another minima, as the energy difference is large. Thus the search will never exit a local minima and simulated annealing won't do much better than gradient descent. On the other hand, making the PAF independent of the difference in energies will allow the algorithm to do its magic.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
````diff
@@ -0,0 +1,198 @@
+- The effect of the difference in energies, $E_{next} - E$, on the PAF can be increased/decreased by increasing/decreasing the base of the exponent as shown below: 
+```cpp
+bool P(double E,double E_next,double T){
+    double e = 2; // set e to any real number greater than 1
````
Ahhh, an engineer wrote this, I see :grin:

adamant-pwn reviewed (changes requested):
Thanks for the PR! I left some comments that I'd like to be addressed before this is merged.

Proxihox commented:
@mhayter @adamant-pwn Any updates in this regard?

mhayter commented:
Thanks for your patience. Unfortunately, for articles where I'm not an expert, I'm leaving a final review to more senior experts like @adamant-pwn and @jakobkogler.

Thank you for creating the article though. I'm sure it will eventually be accepted as it is useful.

mhayter commented:
Is there any update here. This seemed like a worthwhile article to me?

Proxihox commented:
I was caught up with other commitments and I couldn't work on this. I'll send updates over the weekend.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,204 @@
+<br>
+<i>A visual representation of simulated annealing, searching for the maxima of this function with multiple local maxima.</i>.
+<br>
+<i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
```
```suggestion
<i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
</center>
```

Proxihox commented:
Hi @adamant-pwn @mhayter it shows changes requested but I'm not able to see anything. Is there anything I need to do? I'd also like to apologize, I accidently hit close merge request and the preview is no longer visible.

mhayter commented:
I've generated the new preview. If we can't get a review by @adamant-pwn in a week or two, I'll consider publishing it as the original article was useful for me( I was able to piece together a solution using your template.).

Unfortunately we don't have dedicated enough or, perhaps in my case, experienced enough volunteers, so reviews are difficult for new articles.

The lack of timely responses I imagine must be discouraging for current and future authors.

I'm sorry about that.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,205 @@
+    
+    double E() {
+        double dist = 0;
+        bool first = true;
```
```suggestion
```
It seems unused?

adamant-pwn commented on src/num_methods/simulated_annealing.md:
````diff
@@ -0,0 +1,205 @@
+- The effect of the difference in energies, $E_{next} - E$, on the PAF can be increased/decreased by increasing/decreasing the base of the exponent as shown below: 
+```cpp
+bool P(double E, double E_next, double T, mt19937 rng) {
+    e = 2 // set e to any real number greater than 1
+    double prob =  exp(-(E_next-E)/T);
````
```suggestion
    double e = 2; // set e to any real number greater than 1
    double prob =  pow(e, -(E_next-E)/T);
```

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,205 @@
+## Further modifications to the algorithm:
+
+- Add a time based exit condition to the while loop to prevent TLE
+- The decay implemented above is an exponential decay. You can always replace this with a decay function as per your needs.
```
Is there any example where non-exponential decay performs better?

Proxihox replied:
Libraries which implement Sim Anneal usually provide options for non-exponential decay such as [this one](https://mlrose.readthedocs.io/en/stable/source/decay.html). A linear decay function will motivate the algorithm to explore actively even towards the end. This is useful when there are multiple local minima very close to each other.

adamant-pwn reviewed (changes requested):
Hi, thanks for the update! I think we're getting very close here, but there are still some minor suggestion that, I think, should be addressed :slightly_smiling_face:

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,205 @@
+<br>
+<i>A visual representation of simulated annealing, searching for the maxima of this function with multiple local maxima.</i>.
+<br>
+<i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
```
This misrenders in the [GitHub pages preview](https://gh.cp-algorithms.com/1371/num_methods/simulated_annealing.html), but renders correctly at the [built-in preview](https://cp-algorithms.com/preview.html). I'm not really sure what causes this, but, given that the image is under CC0, in principle, we can remove this line altogether.

adamant-pwn commented on src/num_methods/simulated_annealing.md:
```diff
@@ -0,0 +1,205 @@
+<i>This <a href="https://commons.wikimedia.org/wiki/File:Hill_Climbing_with_Simulated_Annealing.gif">animation</a> by [Kingpin13](https://commons.wikimedia.org/wiki/User:Kingpin13) is distributed under <a href="https://creativecommons.org/publicdomain/zero/1.0/deed.en">CC0 1.0</a></i> license.
+</center>
+
+### Temperature($T$) and decay($u$)
```
```suggestion
### Temperature (T) and decay (u)
```
`$` misrenders in sidebar table of contents.

adamant-pwn commented:
Thank you very much for the contribution! And sorry that it had to take a while to do the review :slightly_smiling_face:
